### ISSUE TITLE
Add geodatabase and layer views, accommodate HRUs and MapDocs, and update URLs

### ISSUE THREAD

**Issue by jkeifer** [PSU-CSAR/django-ebagis#24]
I think the views organization may have become a bit unwieldy and contrived. I believe it would make more sense to break everything down into individual views to follow the models that have been defined. That is, have an AOI view, a Geodatabase view, a File view, etc. The special routing that comes from the AOIs having only one of each type of sub-object (i.e., one surfaces, aoidb, analysis, etc) can be accomplished via methods on the AOI view calling the detail view of the respective model view and passing in the appropriate pk value.

Moreover, the new HRU handling has not been fully implemented in the views, serializers, and URLs. The models have been completed allowing full AOI uploads and downloads, but the HRU components are otherwise missing or unfinished.

All of this work may reveal missing entries and/or shortcomings in the URL routing as it is currently defined. The URLs will likely need an overhaul to ensure they are up-to-date and include all the relevant endpoints.

**Comment by jkeifer** [comment]
Note that in addition to the lack of proper HRU handling, MapDoc handling is also not fully implemented. This will require changes/additions in the views, serializers, and URLs, as with the HRUs.

**Comment by jkeifer** [comment]
So after quite a bit of thought, I believe the best way to do this is to follow the pattern  specifically demonstrated in the layers URLs of parameterizing the geodatabase name and feeding that into the get_queryset method for filtering. This pattern can be copied and modified to create a generic endpoint that works for each of the geodatabases. It is likely that a single serializer will work for all of the different geodatabase proxy types. This pattern can also be used to create a generic endpoint of each of the different file types.

Essentially, I believe I will need to create all of the required endpoint URLs first, for all of the different possible scenarios: aoi list, specific aoi, specific geodatabase in a specific aoi, a specific layer type in a geodatabase, etc. Good use of regular expressions here will allow a single URL pattern to match may different situations with a single entry. Then, I will need to figure out what view to point each URL patter at. The get_queryset filtering will take care of selecting only the desired objects.

Each of the views will likely be very generic, simply inheriting the basic ModelViewSet class, overriding the get_queryset method, and adding the download/upload/update mixins as required. It may make sense, from an organizational perspective, to make a base view that includes the get_queryset_arguments method rather than importing it from the utilities.

**Comment by jkeifer** [comment]
To expand upon, "Good use of regular expressions here will allow a single URL pattern to match may different situations with a single entry":

It seems as though most URLs will be structured like this:

```
aois -> aoi_id -> geodatabase/directory_name -> file_type (with an s) -> file_id
```

But we also have a geodatabase/directory endpoint:

```
geodatabase/directory_name (with an s) -> file_type (with an s) -> file_id
```

And of course we have just the file endpoints:

```
file_type (with an s) -> file_id
```

To make it even more complicated, we can also have additional steps in-between, when it comes to PRISM and HRUs, which both have multiple versions, and the latter can support a list of objects:

```
aois -> aoi_id -> prism -> version_id -> file_type (with an s) -> file_id
aois -> aoi_id -> zones -> zones_id -> version_id -> file_type (with an s) -> file_id

prisms -> version_id -> file_type (with an s) -> file_id
zones -> zones_id -> version_id -> file_type (with an s) -> file_id
```

I believe a clear and concise regex exists for these cases that can simplify them into three or so expressions. Such URL simplicity may beget complex view functions. I am undecided if a larger set of regexes should be used to reduce the complexity of the views.

One thought I just had: the structure of the zones hierarchy is due to the fact that zones is a directory object, as is the version "container". Thus, the general-purpose solution here may be to support n-levels of recursion. The get_queryset_arguments method can be used to perform pre-processing of the passed argument names to help enable such a solution.

**Comment by jkeifer** [comment]
So it turns out after giving this (vastly too) much thought, a general case URL regex pattern cannot be constructed. I was aiming for such a solution due to the django-rest-framework `reverse` function, which will return a URL to a model based on a view names as defined in urls.py. This means that one must have a single URL expression mapped to a single view name, and that name is passed to `reverse`. 

However, I have since learned that `reverse` does not work with the `|` operator in the expression, which is used for `or` in the regex language. Because of the multiple possible URLs for any given model, I either have to define unique URL expressions with unique view names for each possibility, or build an overly-complex and unmaintainable regex that uses `|` for conditionals. Either way does not work. Consequentially, I've struggled to find a solution.

I have since realized that `reverse` will only ever work reliably to derive the URL of an object that has no parent objects. This can be thought of like an orphaned raster file: its url would be `rasters/<id>/`. However, if we consider that it has a parent object and we can append a predictable piece to the parent object's URL to get the raster's URL (like `<parent_objs_url>/rasters/<id>/`), then we can derive a recursive process to get the parent object's URL and append that missing piece. The parent object's URL is similar: we do not know how to find the URL directly with `reverse` unless the object is orphaned. And if the object is not orphaned, we can find its URL based on its parent object's URL like before.

Unlike other recursive ideas I had (mentioned in previous comments), the beauty of this approach is that it is based on model instance methods, not model class methods. The problem with the class methods is what I've outlined above: a specific instance has a single parent object and therefore a single URL path, but model classes can have many possible parent objects, so many possible URL path patterns could be needed for its instances.
